File: elixir/lib/symphony_elixir/tools.ex
defmodule SymphonyElixir.Tools do
  @moduledoc """
  Canonical tool registry and execution layer shared across executors.
  """

  alias SymphonyElixir.Linear.Client

  @linear_graphql_tool "linear_graphql"
  @linear_graphql_description """
  Execute a raw GraphQL query or mutation against Linear using Symphony's configured auth.
  """
  @linear_graphql_input_schema %{
    "type" => "object",
    "additionalProperties" => false,
    "required" => ["query"],
    "properties" => %{
      "query" => %{
        "type" => "string",
        "description" => "GraphQL query or mutation document to execute against Linear."
      },
      "variables" => %{
        "type" => ["object", "null"],
        "description" => "Optional GraphQL variables object.",
        "additionalProperties" => true
      }
    }
  }

  @type execution_result ::
          {:ok, %{success: boolean(), payload: term()}} | {:error, map()}

  @spec tool_specs() :: [map()]
  def tool_specs do
    [
      %{
        "name" => @linear_graphql_tool,
        "description" => @linear_graphql_description,
        "inputSchema" => @linear_graphql_input_schema
      }
    ]
  end

  @spec supported_tool_names() :: [String.t()]
  def supported_tool_names do
    Enum.map(tool_specs(), & &1["name"])
  end

  @spec execute(String.t() | nil, term(), keyword()) :: execution_result()
  def execute(tool, arguments, opts \\ []) do
    do_execute(tool, arguments, opts)
  end

  defp do_execute(@linear_graphql_tool, arguments, opts) do
    case normalize_linear_graphql_arguments(arguments) do
      {:ok, query, variables} ->
        linear_client = Keyword.get(opts, :linear_client, &Client.graphql/3)

        case linear_client.(query, variables, []) do
          {:ok, response} ->
            {:ok, %{success: graphql_success?(response), payload: response}}

          {:error, reason} ->
            {:error, tool_error_payload(reason)}
        end

      {:error, reason} ->
        {:error, tool_error_payload(reason)}
    end
  end

  defp do_execute(_tool, _arguments, _opts), do: {:error, unsupported_tool_payload()}

  @spec unsupported_tool_payload() :: map()
  def unsupported_tool_payload do
    %{
      "error" => %{
        "message" => "Unsupported tool.",
        "supportedTools" => supported_tool_names()
      }
    }
  end

  @spec encode_payload(term()) :: String.t()
  def encode_payload(payload) when is_map(payload) or is_list(payload) do
    Jason.encode!(payload, pretty: true)
  end

  def encode_payload(payload), do: inspect(payload)

  defp normalize_linear_graphql_arguments(arguments) when is_binary(arguments) do
    case String.trim(arguments) do
      "" -> {:error, :missing_query}
      query -> {:ok, query, %{}}
    end
  end

  defp normalize_linear_graphql_arguments(arguments) when is_map(arguments) do
    case normalize_query(arguments) do
      {:ok, query} ->
        case normalize_variables(arguments) do
          {:ok, variables} -> {:ok, query, variables}
          {:error, reason} -> {:error, reason}
        end

      {:error, reason} ->
        {:error, reason}
    end
  end

  defp normalize_linear_graphql_arguments(_arguments), do: {:error, :invalid_arguments}

  defp normalize_query(arguments) do
    case Map.get(arguments, "query") || Map.get(arguments, :query) do
      query when is_binary(query) ->
        case String.trim(query) do
          "" -> {:error, :missing_query}
          trimmed -> {:ok, trimmed}
        end

      _ ->
        {:error, :missing_query}
    end
  end

  defp normalize_variables(arguments) do
    case Map.get(arguments, "variables") || Map.get(arguments, :variables) || %{} do
      variables when is_map(variables) -> {:ok, variables}
      _ -> {:error, :invalid_variables}
    end
  end

  defp graphql_success?(response) do
    case response do
      %{"errors" => errors} when is_list(errors) and errors != [] -> false
      %{errors: errors} when is_list(errors) and errors != [] -> false
      _ -> true
    end
  end

  defp tool_error_payload(:missing_query) do
    %{
      "error" => %{
        "message" => "`linear_graphql` requires a non-empty `query` string."
      }
    }
  end

  defp tool_error_payload(:invalid_arguments) do
    %{
      "error" => %{
        "message" => "`linear_graphql` expects either a GraphQL query string or an object with `query` and optional `variables`."
      }
    }
  end

  defp tool_error_payload(:invalid_variables) do
    %{
      "error" => %{
        "message" => "`linear_graphql.variables` must be a JSON object when provided."
      }
    }
  end

  defp tool_error_payload(:missing_linear_api_token) do
    %{
      "error" => %{
        "message" => "Symphony is missing Linear auth. Set `linear.api_key` in `WORKFLOW.md` or export `LINEAR_API_KEY`."
      }
    }
  end

  defp tool_error_payload({:linear_api_status, status}) do
    %{
      "error" => %{
        "message" => "Linear GraphQL request failed with HTTP #{status}.",
        "status" => status
      }
    }
  end

  defp tool_error_payload({:linear_api_request, reason}) do
    %{
      "error" => %{
        "message" => "Linear GraphQL request failed before receiving a successful response.",
        "reason" => inspect(reason)
      }
    }
  end

  defp tool_error_payload(reason) do
    %{
      "error" => %{
        "message" => "Linear GraphQL tool execution failed.",
        "reason" => inspect(reason)
      }
    }
  end
end

File: ts/extensions/linear-tracker/src/tools.ts
import path from "node:path";
import { fileURLToPath } from "node:url";

import { errorMessage, isRecord, type Settings } from "@lorenz/domain";
import {
  toolFailure,
  toolSuccess,
  unsupportedToolFailure,
  type ToolProvider,
  type ToolResult,
  type ToolSpec,
} from "@lorenz/tool-sdk";

import { linearToolPackOptions, validateLinearToolOptions } from "./options.js";

/**
 * The `symphony-linear` skill ships inside this extension (`skills/symphony-linear`, a sibling
 * of `src`/`dist`) so mounting the Linear pack also overlays the doc that teaches the agent to
 * call `linear_graphql`. Resolved from this module so it works from both `src` and `dist`.
 */
const symphonyLinearSkillDir = path.resolve(
  path.dirname(fileURLToPath(import.meta.url)),
  "..",
  "skills",
  "symphony-linear",
);

const LINEAR_MAX_RETRIES = 4;
const MAX_ERROR_BODY_LOG_BYTES = 1000;

interface LinearToolLogger {
  warn(message: string): void;
  error(message: string): void;
}

const defaultLinearToolLogger: LinearToolLogger = {
  warn: (message) => console.warn(message),
  error: (message) => console.error(message),
};

export function linearToolSpecs(): ToolSpec[] {
  return [
    {
      name: "linear_graphql",
      description: "Run a Linear GraphQL operation using Symphony's configured Linear credentials.",
      inputSchema: {
        type: "object",
        additionalProperties: false,
        properties: {
          query: {
            type: "string",
            description: "GraphQL query or mutation document to execute against Linear.",
          },
          variables: {
            type: ["object", "null"],
            description: "Optional GraphQL variables object.",
            additionalProperties: true,
          },
        },
        required: ["query"],
      },
    },
  ];
}

export async function executeLinearTool(
  name: string,
  input: unknown,
  settings: Settings,
  fetchImpl: typeof fetch = fetch,
): Promise<ToolResult> {
  if (name !== "linear_graphql") {
    return unsupportedToolFailure(name, ["linear_graphql"]);
  }
  const normalizedInput = normalizeLinearGraphqlInput(input);
  if (!normalizedInput.ok) {
    return toolFailure(normalizedInput.error);
  }
  const { apiKey, endpoint } = linearToolPackOptions(settings);
  if (!apiKey)
    return toolFailure(
      "Symphony is missing Linear auth. Set `linear.api_key` in `WORKFLOW.md` or export `LINEAR_API_KEY`.",
    );
  if (
    normalizedInput.variables !== undefined &&
    normalizedInput.variables !== null &&
    !isRecord(normalizedInput.variables)
  ) {
    return toolFailure("`linear_graphql.variables` must be a JSON object when provided.");
  }

  try {
    const logger = defaultLinearToolLogger;
    const response = await fetchWithRateLimitRetry(
      fetchImpl,
      { apiKey, endpoint },
      normalizedInput.query,
      normalizedInput.variables ?? {},
      logger,
    );
    const bodyResult = await readResponseJson(response);
    if (!bodyResult.ok) {
      if (!response.ok) {
        logStatusError(logger, normalizedInput.query, response.status, bodyResult.rawBody);
        return toolFailure(`Linear GraphQL request failed with HTTP ${response.status}.`, {
          status: response.status,
        });
      }
      return toolFailure(`linear_invalid_json: ${errorMessage(bodyResult.error)}`);
    }
    const body = bodyResult.body;
    if (!response.ok) {
      logStatusError(logger, normalizedInput.query, response.status, body);
      return toolFailure(`Linear GraphQL request failed with HTTP ${response.status}.`, {
        status: response.status,
      });
    }
    if (isRecord(body) && Array.isArray(body.errors) && body.errors.length > 0) {
      return { success: false, result: body };
    }
    return toolSuccess(body);
  } catch (error) {
    defaultLinearToolLogger.error(
      `Linear GraphQL request failed: ${errorMessage(error)}${linearErrorContext(normalizedInput.query)}`,
    );
    return toolFailure("Linear GraphQL request failed before receiving a successful response.", {
      reason: errorMessage(error),
    });
  }
}

/** The Linear tool pack: raw GraphQL access using the pack's own Linear credentials. */
export const linearToolProvider: ToolProvider = {
  name: "linear",
  skills: [symphonyLinearSkillDir],
  validateOptions: validateLinearToolOptions,
  toolSpecs: () => linearToolSpecs(),
  executeTool: async (name, input, context) =>
    executeLinearTool(name, input, context.settings, context.fetchImpl),
};

function normalizeLinearGraphqlInput(
  input: unknown,
):
  | { ok: true; query: string; variables?: Record<string, unknown> | null | undefined }
  | { ok: false; error: string } {
  if (typeof input === "string") {
    if (input.trim() === "")
      return { ok: false, error: "`linear_graphql` requires a non-empty `query` string." };
    return { ok: true, query: input };
  }
  if (!isRecord(input) || typeof input.query !== "string") {
    return { ok: false, error: "`linear_graphql` requires a non-empty `query` string." };
  }
  if (input.query.trim() === "")
    return { ok: false, error: "`linear_graphql` requires a non-empty `query` string." };
  if (input.variables !== undefined && input.variables !== null && !isRecord(input.variables)) {
    return { ok: false, error: "`linear_graphql.variables` must be a JSON object when provided." };
  }
  return {
    ok: true,
    query: input.query,
    variables: input.variables as Record<string, unknown> | null,
  };
}

async function fetchWithRateLimitRetry(
  fetchImpl: typeof fetch,
  auth: { apiKey: string; endpoint: string },
  query: string,
  variables: Record<string, unknown>,
  logger: LinearToolLogger,
): Promise<Response> {
  for (let retryCount = 0; ; retryCount += 1) {
    const response = await fetchImpl(auth.endpoint, {
      method: "POST",
      signal: AbortSignal.timeout(30_000),
      headers: {
        "content-type": "application/json",
        authorization: auth.apiKey,
      },
      body: JSON.stringify({ query, variables }),
    });
    if (response.status !== 429 || retryCount >= LINEAR_MAX_RETRIES) return response;
    const delayMs = retryDelayMs(response.headers, retryCount);
    logger.warn(
      `Linear GraphQL request rate limited status=429 retry=${retryCount + 1}/${LINEAR_MAX_RETRIES} delay_ms=${delayMs}${linearErrorContext(query, await responseBodyForLog(response))}`,
    );
    await sleep(delayMs);
  }
}

function retryDelayMs(headers: Headers, retryCount: number): number {
  const retryAfter = headers.get("retry-after");
  if (retryAfter !== null) {
    const seconds = Number(retryAfter.trim());
    if (Number.isInteger(seconds) && seconds >= 0) return seconds * 1000;
    const dateMs = Date.parse(retryAfter.trim());
    if (!Number.isNaN(dateMs)) return Math.max(0, dateMs - Date.now());
  }
  return Math.min(1_000 * 2 ** retryCount, 30_000);
}

async function sleep(delayMs: number): Promise<void> {
  await new Promise((resolve) => setTimeout(resolve, delayMs));
}

async function readResponseJson(
  response: Response,
): Promise<{ ok: true; body: unknown } | { ok: false; error: unknown; rawBody?: string }> {
  let rawBody: string;
  try {
    rawBody = await response.text();
  } catch (error) {
    return { ok: false, error };
  }
  try {
    return { ok: true, body: JSON.parse(rawBody) as unknown };
  } catch (error) {
    return { ok: false, error, rawBody };
  }
}

async function responseBodyForLog(response: Response): Promise<unknown> {
  try {
    const rawBody = await response.clone().text();
    try {
      return JSON.parse(rawBody) as unknown;
    } catch {
      return rawBody;
    }
  } catch {
    return undefined;
  }
}

function logStatusError(
  logger: LinearToolLogger,
  query: string,
  status: number,
  body: unknown,
): void {
  logger.error(`Linear GraphQL request failed status=${status}${linearErrorContext(query, body)}`);
}

function linearErrorContext(query: string, body?: unknown): string {
  const parts: string[] = [];
  const operation = operationName(query);
  if (operation) parts.push(`operation=${operation}`);
  if (body !== undefined) parts.push(`body=${summarizeErrorBody(body)}`);
  return parts.length === 0 ? "" : ` ${parts.join(" ")}`;
}

function operationName(query: string): string | null {
  return /\b(?:query|mutation)\s+([A-Za-z_][A-Za-z0-9_]*)/.exec(query)?.[1] ?? null;
}

function summarizeErrorBody(body: unknown): string {
  const text = typeof body === "string" ? body : stringifyErrorBody(body);
  const compact = text.replace(/\s+/g, " ").trim();
  if (compact.length <= MAX_ERROR_BODY_LOG_BYTES) return compact;
  return `${compact.slice(0, MAX_ERROR_BODY_LOG_BYTES)}...<truncated>`;
}

function stringifyErrorBody(body: unknown): string {
  try {
    return JSON.stringify(body) ?? String(body);
  } catch {
    return String(body);
  }
}
